core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -212,6 +226,70 @@ class half_t {
   KOKKOS_FUNCTION
   half_t(unsigned long long rhs) : val(cast_to_half(rhs).val) {}
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void atomic_add(const volatile half_t src) const volatile {
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    volatile float* lhs = (float*)&val;
+    volatile float* rhs = (float*)&src.val;
+    *lhs                = __half2float(val);
+    *rhs                = __half2float(src.val);
+    val                 = __float2half(lhs[0] + rhs[0]);
+#else
+    // Cuda 10 supports __half volatile stores but not volatile arithmetic
+    // operands so we use atomicAdd for cuda 10.0 instead.
```

Comment:
Cuda 10 introduced atomics for `__half` (actually there were assembly instructions for atomics in Cuda 9) so this will work for cuda 10 or greater. Originally, I thought we could simply perform the arithmetic without volatile operands (since volatile arithmetic is not supported in the `__half` type) and then do a volatile store with the arithmetic result; the line of thought was that if every store is volatile then it will never get placed in a cache along the way so that when the value is loaded, it will effectively be a volatile load since the hardware will only find the data all the way out in memory. When testing this, I found race conditions in the reduction code; this is why I am using __half atomics now.